src/cas_cli/client.py

```diff
@@ -0,0 +1,238 @@
+import asyncio
+import datetime
+import functools
+import math
+import operator
+import os
+import time
+import typing as t
+
+import anndata
+
+from cas_cli import _read_data, data_preparation, exceptions, service
+
+
+class CASClient:
+    """
+    Service that is designed to call Cellarium Cloud Backend API service
+    """
+
+    def __init__(self, api_token: str) -> None:
+        self._print("Connecting with Cellarium Cloud backend...")
+        self.cas_api_service = service.CASAPIService(api_token=api_token)
+        self.feature_schemas = []
+        self._feature_schemas_cache = {}
+
+    @staticmethod
+    def _get_number_of_chunks(adata, chunk_size):
+        return math.ceil(len(adata) / chunk_size)
+
+    @staticmethod
+    def _get_timestamp() -> str:
+        return datetime.datetime.now().strftime("%H:%M:%S.%f")[:-3]
+
+    def _print(self, str_to_print: str) -> None:
+        print(f"* [{self._get_timestamp()}] {str_to_print}")
+
+    def _validate_and_sanitize_input_data(
+        self,
+        adata: anndata.AnnData,
+        feature_schema_name: str,
+        count_matrix_name: str,
+        feature_ids_column_name: str,
+    ) -> anndata.AnnData:
+        try:
+            cas_feature_schema_list = self._feature_schemas_cache[feature_schema_name]
+        except KeyError:
+            cas_feature_schema_list = self.cas_api_service.get_cas_pca_002_schema_from_dump()
+            self._feature_schemas_cache[feature_schema_name] = cas_feature_schema_list
+
+        try:
+            data_preparation.validate(
+                adata=adata,
+                cas_feature_schema_list=cas_feature_schema_list,
+                feature_ids_column_name=feature_ids_column_name,
+            )
+        except exceptions.DataValidationError as e:
+            if e.extra_features > 0:
+                self._print(
+                    f"Input datafile has {e.extra_features} more features than {feature_schema_name} schema in CAS. "
+                    f"They will be omitted..."
+                )
+            if e.missing_features > 0:
+                self._print(
+                    f"Input datafile has {e.missing_features} missing features compared with {feature_schema_name} in CAS. "
+                    f"They will be filled with zeros..."
+                )
+            if e.extra_features == 0 and e.missing_features == 0:
+                self._print(
+                    f"Input datafile has all the necessary features as {feature_schema_name}, but it's still "
```

Comment:
f'The input data matrix contains all of the features specified in "{feature_schema_name}" CAS schema but '
f'in a different order. The input features will be reordered according to "{feature_schema_name}".'